teps/0145-cel-in-whenexpression.md

```diff
@@ -0,0 +1,339 @@
+- Support numeric operators such as GreaterThan/LessThan, e.g. Execute the task if test coverage is large than a certain number.
+
+### Goals
+- Support CEL in `WhenExpression`.
+
+### Non-Goals
+- Support CEL in Pipeline API fields other than `WhenExpression`.
+
+### Related Work
+
+Similar expression evaluations are supported in other CI/CD tools in the industry:
+- [Argo][Argo]: `Workflows` support conditional execution using a when property to specify whether to run or skip a `Step` ([example][argo example]).
+- [GitHub Actions][GitHub Actions]: supports evaluating expressions in `if` of a `step` to enable conditional execution.
+- [Jenkins][Jenkins]: executes the stage when the specified Groovy expression evaluates to true.
+- [Spinnaker][Spinnaker]: Uses string expressions, `Stages` only run when expressions evaluate to `True`.
+
+
+## Proposal
+
+This doc proposes to add a `string` type field `cel` under `WhenExpression`. Like [CelCustomRun][CELCustomRun], we propose to use [cel-go][cel-go] for the expression evaluation. This will allow users to craft any valid CEL expression as defined by the [cel-spec][cel-spec] language definition.
+
+
+### Syntax
+
+In this section we will show how CEL can support the [use cases](#use-cases).
+
+#### Use CEL to replace current WhenExpression
+```yaml
+# current WhenWxpressions
+when:
+  - input: "foo"
+    operator: "in"
+    values: ["foo", “bar”]
+  - input: "duh"
+    operator: "notin"
+    values: ["foo", “bar”]
+
+# with cel
+when:
+  - cel: "\”foo\” in [\"foo\", \“bar\”]”
+  - cel: "\”duh\” not in [\"foo\", \“bar\”]”
+```
+
+**Note:** The `WhenExpressions` contains a list of `WhenExpression`, if all `WhenExpressions` are true it will return true and allow execution. So we’re using AND to evaluate the list of `WhenExpressions`. And this will remain the same for this proposal. Each existing `WhenExpression` can be replaced with a CEL expression.
+
+#### Use CEL to Support ANY operator
+
+This is the workaround from [#3591][#3591],
+```yaml
+  when:
+    - input: "$(params.param1)$(params.param2)"
+      operator: notin
+      values: [""]
+```
+
+With CEL we can use a more intuitive expression to support it.
+```yaml
+  when:
+    - cel: "\"$(params.param1)\" != \"\" || \"$(params.param2)\" != \"\""
+```
+
+#### Use CEL to Support Numeric Comparisons
+
+This is not possible with current `WhenExpression`. With CEL we can do:
+
+```yaml
+  when:
+    - cel: "\"$(tasks.unit-test.results.test-coverage)\" >= 0.9"
+```
+
+[Note:][numeric note] currently there are no automatic arithmetic conversions for the numeric types (int, uint, and double), so the emitted results must be the same numeric type to make sure the evaluation works.
+
+
+### Validation
+
+The new fields will be gated by `enable-api-fields`, which will default to false while the feature is in alpha.When per feature flag (TEP-0138) is implemented, we will use a dedicated feature flag `enable-cel-when-expression` to gate this feature.
```

Comment:
Why do we talk about `enable-api-fields` at all here ? We can just decide to only use a dedicated feature flag.

Reply:
Yes my previous version of the proposal is using a dedicated feature flag. But I saw the comment here https://github.com/tektoncd/pipeline/pull/7162#discussion_r1342965164 and changed it to the group flag. 
@JeromeJu  Would a dedicated feature flag be ok to use?

Reply:
Thanks for the checkin!
Currently we are able to/ should switch to per-feature flag! But while actually making that happen we've been waiting for some of the PRs to get merged in:
i.e. https://github.com/tektoncd/pipeline/pull/6941 and another [setup for the CI tests](https://github.com/tektoncd/plumbing/issues/1629) (not blocking any per-feature flag but would be nice to have the plumbing set up).

Reply:
I will update this proposal to use dedicated feature flag

Reply:
Thanks @Yongxuanzhang for the responsive conversation and updates!